The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10588373997/job/29340786168) [(plain)](https://github.com/rust-lang/rust/commit/13e36cf8870a88ab0ac5651d9513b13085baa8cc/checks/29340786168/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.737 Building wheels for collected packages: reuse
#13 2.738   Building wheel for reuse (pyproject.toml): started
#13 2.989   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 2.990   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 2.990   Stored in directory: /tmp/pip-ephem-wheel-cache-ps2byme7/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 2.993 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.394 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.394 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.931 Collecting virtualenv
#13 3.931 Collecting virtualenv
#13 3.971   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.044      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 80.7 MB/s eta 0:00:00
#13 4.084 Collecting distlib<1,>=0.3.7
#13 4.089   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.098      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 70.4 MB/s eta 0:00:00
#13 4.129 Collecting platformdirs<5,>=3.9.1
#13 4.133   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.168 Collecting filelock<4,>=3.12.2
#13 4.172   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.256 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.443 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      204736 kB
DirectMap2M:     6086656 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/1f12b9b0fdbe735968ac002792a720f0ba4faca6/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-1f12b9b0fdbe735968ac002792a720f0ba4faca6-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
##[endgroup]
fmt check
Diff in /checkout/library/alloc/src/rc.rs:1710:
     #[inline]
     #[stable(feature = "rc_unique", since = "1.4.0")]
     pub fn get_mut(this: &mut Self) -> Option<&mut T> {
-        if Rc::is_unique(this) {
-            unsafe { Some(Rc::get_mut_unchecked(this)) }
-            None
-        }
-        }
+        if Rc::is_unique(this) { unsafe { Some(Rc::get_mut_unchecked(this)) } } else { None }
 
     /// Returns a mutable reference into the given `Rc`,
Diff in /checkout/library/alloc/src/rc.rs:3291:
     #[must_use]
     #[must_use]
     #[stable(feature = "weak_counts", since = "1.41.0")]
     pub fn strong_count(&self) -> usize {
-        if let Some(inner) = self.inner() {
-            inner.strong()
-        } else {
-            0
-        }
+        if let Some(inner) = self.inner() { inner.strong() } else { 0 }
 
 
     /// Gets the number of `Weak` pointers pointing to this allocation.
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/librustdoc/clean/render_macro_matchers.rs" "/checkout/src/librustdoc/clean/utils.rs" "/checkout/src/librustdoc/clean/utils/tests.rs" "/checkout/src/librustdoc/clean/blanket_impl.rs" "/checkout/src/librustdoc/clean/mod.rs" "/checkout/src/librustdoc/clean/types.rs" "/checkout/src/librustdoc/clean/auto_trait.rs" "/checkout/src/librustdoc/clean/simplify.rs" "/checkout/src/librustdoc/clean/inline.rs" "/checkout/src/librustdoc/clean/cfg/tests.rs" "/checkout/src/librustdoc/clean/types/tests.rs" "/checkout/src/librustdoc/clean/cfg.rs" "/checkout/src/librustdoc/scrape_examples.rs" "/checkout/library/alloc/src/macros.rs" "/checkout/library/alloc/src/alloc/tests.rs" "/checkout/library/alloc/src/sync/tests.rs" "/checkout/library/alloc/src/slice/tests.rs" "/checkout/library/alloc/src/rc.rs" "/checkout/src/librustdoc/html/length_limit.rs" "/checkout/src/librustdoc/html/toc.rs" "/checkout/src/librustdoc/html/highlight.rs" "/checkout/src/librustdoc/html/layout.rs" "/checkout/src/librustdoc/html/length_limit/tests.rs" "/checkout/src/librustdoc/html/toc/tests.rs" "/checkout/src/librustdoc/html/mod.rs" "/checkout/library/alloc/src/vec/spec_from_iter.rs" "/checkout/library/alloc/src/vec/splice.rs" "/checkout/library/alloc/src/vec/in_place_collect.rs" "/checkout/library/alloc/src/vec/mod.rs" "/checkout/library/alloc/src/vec/partial_eq.rs" "/checkout/library/alloc/src/vec/extract_if.rs" "/checkout/library/alloc/src/vec/set_len_on_drop.rs" "/checkout/library/alloc/src/vec/spec_extend.rs" "/checkout/library/alloc/src/vec/in_place_drop.rs" "/checkout/library/alloc/src/vec/drain.rs" "/checkout/library/alloc/src/vec/cow.rs" "/checkout/library/alloc/src/vec/is_zero.rs" "/checkout/library/alloc/src/vec/spec_from_iter_nested.rs" "/checkout/library/alloc/src/vec/into_iter.rs" "/checkout/library/alloc/src/vec/spec_from_elem.rs" "/checkout/library/alloc/src/raw_vec.rs" "/checkout/library/alloc/src/sync.rs" "/checkout/library/alloc/src/lib.miri.rs" "/checkout/library/alloc/src/lib.rs" "/checkout/library/alloc/src/alloc.rs" "/checkout/library/alloc/src/boxed.rs" "/checkout/src/librustdoc/html/escape/tests.rs" "/checkout/src/librustdoc/html/url_parts_builder/tests.rs" "/checkout/src/librustdoc/html/markdown.rs" "/checkout/src/librustdoc/html/markdown/tests.rs" "/checkout/src/librustdoc/html/static_files.rs" "/checkout/src/librustdoc/html/highlight/fixtures/sample.rs" "/checkout/src/librustdoc/html/highlight/fixtures/union.rs" "/checkout/src/librustdoc/html/highlight/tests.rs" "/checkout/src/librustdoc/html/url_parts_builder.rs" "/checkout/library/alloc/src/collections/btree/fix.rs" "/checkout/library/alloc/src/collections/btree/set/tests.rs" "/checkout/library/alloc/src/collections/btree/map.rs" "/checkout/library/alloc/src/collections/btree/set.rs" "/checkout/library/alloc/src/collections/btree/mod.rs" "/checkout/library/alloc/src/collections/btree/node/tests.rs" "/checkout/library/alloc/src/collections/btree/set_val.rs" "/checkout/library/alloc/src/collections/btree/append.rs" "/checkout/library/alloc/benches/binary_heap.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Wed Aug 28 01:16:02 UTC 2024
  network time: Wed, 28 Aug 2024 01:16:02 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/library/alloc/src/rc.rs
+++ b/library/alloc/src/rc.rs
@@ -687,54 +687,6 @@ impl<T, A: Allocator> Rc<T, A> {
         }
     }
 
-    /// TODO: document
-    #[cfg(not(no_global_oom_handling))]
-    #[unstable(feature = "allocator_api", issue = "32838")]
-    pub fn new_cyclic_in<F>(data_fn: F, alloc: A) -> Rc<T, A>
-    where
-        F: FnOnce(&Weak<T, A>) -> T,
-    {
-        // Construct the inner in the "uninitialized" state with a single
-        // weak reference.
-        let uninit_ptr: NonNull<_> = Box::leak(
-            Box::new_in(RcBox {
-                strong: Cell::new(0),
-                weak: Cell::new(1),
-                value: mem::MaybeUninit::<T>::uninit(),
-            }),
-            alloc,
-        )
-        .into();
-
-        let init_ptr: NonNull<RcBox<T>> = uninit_ptr.cast();
-
-        let weak = Weak { ptr: init_ptr, alloc: Global };
-
-        // It's important we don't give up ownership of the weak pointer, or
-        // else the memory might be freed by the time `data_fn` returns. If
-        // we really wanted to pass ownership, we could create an additional
-        // weak pointer for ourselves, but this would result in additional
-        // updates to the weak reference count which might not be necessary
-        // otherwise.
-        let data = data_fn(&weak);
-
-        let strong = unsafe {
-            let inner = init_ptr.as_ptr();
-            ptr::write(ptr::addr_of_mut!((*inner).value), data);
-
-            let prev_value = (*inner).strong.get();
-            debug_assert_eq!(prev_value, 0, "No prior strong references should exist");
-            (*inner).strong.set(1);
-
-            Rc::from_inner(init_ptr)
-        };
-
-        // Strong references should collectively own a shared weak reference,
-        // so don't run the destructor for our old weak reference.
-        mem::forget(weak);
-        strong
-    }
-
     /// Constructs a new `Rc` with uninitialized contents in the provided allocator.
     ///
     /// # Examples
@@ -2309,6 +2261,85 @@ impl<T: ?Sized, A: Allocator + Clone> Clone for Rc<T, A> {
             self.inner().inc_strong();
             Self::from_inner_in(self.ptr, self.alloc.clone())
         }
+    }
+}
+
+impl<T, A: Allocator + Clone> Rc<T, A> {
+    /// Constructs a new `Rc<T, A>` in the given allocator while giving you a `Weak<T, A>` to the allocation,
+    /// to allow you to construct a `T` which holds a weak pointer to itself.
+    ///
+    /// Generally, a structure circularly referencing itself, either directly or
+    /// indirectly, should not hold a strong reference to itself to prevent a memory leak.
+    /// Using this function, you get access to the weak pointer during the
+    /// initialization of `T`, before the `Rc<T, A>` is created, such that you can
+    /// clone and store it inside the `T`.
+    ///
+    /// `new_cyclic` first allocates the managed allocation for the `Rc<T, A>`,
+    /// then calls your closure, giving it a `Weak<T, A>` to this allocation,
+    /// and only afterwards completes the construction of the `Rc<T, A>` by placing
+    /// the `T` returned from your closure into the allocation.
+    ///
+    /// Since the new `Rc<T, A>` is not fully-constructed until `Rc<T, A>::new_cyclic`
+    /// returns, calling [`upgrade`] on the weak reference inside your closure will
+    /// fail and result in a `None` value.
+    ///
+    /// # Panics
+    ///
+    /// If `data_fn` panics, the panic is propagated to the caller, and the
+    /// temporary [`Weak<T, A>`] is dropped normally.
+    ///
+    /// # Examples
+    /// See [`new_cyclic`].
+    ///
+    /// [`new_cyclic`]: Rc::new_cyclic
+    /// [`upgrade`]: Weak::upgrade
+    #[cfg(not(no_global_oom_handling))]
+    #[unstable(feature = "allocator_api", issue = "32838")]
+    pub fn new_cyclic_in<F>(data_fn: F, alloc: A) -> Rc<T, A>
+    where
+        F: FnOnce(&Weak<T, A>) -> T,
+    {
+        // Note: comments and implementation are copied from Rc::new_cyclic.
+
+        // Construct the inner in the "uninitialized" state with a single
+        // weak reference.
+        let uninit_ptr: NonNull<_> = Box::leak(Box::new_in(
+            RcBox {
+                strong: Cell::new(0),
+                weak: Cell::new(1),
+                value: mem::MaybeUninit::<T>::uninit(),
+            },
+            alloc.clone(),
+        ))
+        .into();
+
+        let init_ptr: NonNull<RcBox<T>> = uninit_ptr.cast();
+
+        let weak = Weak { ptr: init_ptr, alloc: alloc.clone() };
+
+        // It's important we don't give up ownership of the weak pointer, or
+        // else the memory might be freed by the time `data_fn` returns. If
+        // we really wanted to pass ownership, we could create an additional
+        // weak pointer for ourselves, but this would result in additional
+        // updates to the weak reference count which might not be necessary
+        // otherwise.
+        let data = data_fn(&weak);
+
+        let strong = unsafe {
+            let inner = init_ptr.as_ptr();
+            ptr::write(ptr::addr_of_mut!((*inner).value), data);
+
+            let prev_value = (*inner).strong.get();
+            debug_assert_eq!(prev_value, 0, "No prior strong references should exist");
+            (*inner).strong.set(1);
+
+            Rc::from_inner_in(init_ptr, alloc)
+        };
+
+        // Strong references should collectively own a shared weak reference,
+        // so don't run the destructor for our old weak reference.
+        mem::forget(weak);
+        strong
     }
 }
 

--- a/library/alloc/src/rc.rs
+++ b/library/alloc/src/rc.rs
@@ -1710,11 +1710,7 @@ impl<T: ?Sized, A: Allocator> Rc<T, A> {
     #[inline]
     #[stable(feature = "rc_unique", since = "1.4.0")]
     pub fn get_mut(this: &mut Self) -> Option<&mut T> {
-        if Rc::is_unique(this) {
-            unsafe { Some(Rc::get_mut_unchecked(this)) }
-        } else {
-            None
-        }
+        if Rc::is_unique(this) { unsafe { Some(Rc::get_mut_unchecked(this)) } } else { None }
     }
 
     /// Returns a mutable reference into the given `Rc`,
@@ -3291,11 +3287,7 @@ impl<T: ?Sized, A: Allocator> Weak<T, A> {
     #[must_use]
     #[stable(feature = "weak_counts", since = "1.41.0")]
     pub fn strong_count(&self) -> usize {
-        if let Some(inner) = self.inner() {
-            inner.strong()
-        } else {
-            0
-        }
+        if let Some(inner) = self.inner() { inner.strong() } else { 0 }
     }
 
     /// Gets the number of `Weak` pointers pointing to this allocation.

--- a/library/alloc/src/rc.rs
+++ b/library/alloc/src/rc.rs
@@ -2279,7 +2279,7 @@ impl<T, A: Allocator + Clone> Rc<T, A> {
     /// and only afterwards completes the construction of the `Rc<T, A>` by placing
     /// the `T` returned from your closure into the allocation.
     ///
-    /// Since the new `Rc<T, A>` is not fully-constructed until `Rc<T, A>::new_cyclic`
+    /// Since the new `Rc<T, A>` is not fully-constructed until `Rc<T, A>::new_cyclic_in`
     /// returns, calling [`upgrade`] on the weak reference inside your closure will
     /// fail and result in a `None` value.
     ///
@@ -2289,6 +2289,7 @@ impl<T, A: Allocator + Clone> Rc<T, A> {
     /// temporary [`Weak<T, A>`] is dropped normally.
     ///
     /// # Examples
+    ///
     /// See [`new_cyclic`].
     ///
     /// [`new_cyclic`]: Rc::new_cyclic

